### DISCUSSION TITLE
Upgrade fails on Synology

### DISCUSSION THREAD

**Discussion by billraff** [jasonacox/Powerwall-Dashboard#459] [Q&A] [answered]
I wanted to upgrade from 4.0.3 to 4.2.0 and I saw in the release notes it was to remove compatibility with Docker V1. The error I receive is that my Docker Compose is the V1 version. But when I look at the version I see -
<img width="528" alt="Screenshot 2024-03-26 at 10 24 19 AM" src="https://github.com/jasonacox/Powerwall-Dashboard/assets/29167787/f24df09e-424b-46ab-9c31-57472cfccf00">

Anyone familiar with Synology know what is going on?

**Comment by BJReplay** [comment]
Hi @billraff, not familiar with Synology, but @jasonacox and I discussed this before making the change; we mentioned Synology as a possible problem but weren't able to test.  It looks like we will have to roll this change back, unless (or until) we work out if Synology is stuck with V1 docker.

The discussions are here for issue #453 and in PR #454 

I'll have a look and see if there are any options for upgrading docker on synology - as docker V1 is very old: 2016 :)

Now that I have a closer look at your dashboard, I see that you've run `docker-compose --version`.

Can you run `docker compose version` and tell me what you see, if anything?

It looks like you might have a newer version of docker compose, but one that only runs with the docker-compose format rather than docker compose.  If so, we'll need to work out how to fix the test in upgrade.sh

**Reply by jasonacox** [reply]
> docker-compose --version
> Docker Compose version v2.9.0-6413-g38f6acd

Hum.. if this version works with the V2 format, I think we should adjust `upgrade.sh` to allow the `docker-compose` command but detect the version. It will require a bit more logic but is possible.

@billraff Can you try something for us?
* edit the `powerwall.yml` file and remove the `version` line
* run `./compose-dash.sh up -d` and report results

**Reply by jasonacox** [reply]
@BJReplay I wonder if this would work:

```bash
echo "Checking Docker Compose..."
if docker compose version > /dev/null 2>&1; then
    # Build Docker (v2)
    echo "INFO: Docker Compose V2 Found: OK to Continue"
    echo ""
else
    if docker-compose version > /dev/null 2>&1; then
        # Check for version information to see if it is >= 2.0.0
        COMPOSE_VERSION=`docker-compose version --short`
        if [[ "${COMPOSE_VERSION}" == "2"* ]]; then
            # Build Docker (v2)
            echo "INFO: Docker Compose V2 Found: OK to Continue"
        else
            # Build Docker (v1)
            echo "ERROR: Docker Compose V1 Found: Upgrade Required"
            echo "See Migration Instructions at https://docs.docker.com/compose/migrate/"
            exit 1
        fi
    else
        echo "ERROR: Docker Compose is not available."
        echo "This script requires Docker Compose."
        echo "Please install and try again."
    fi
    exit 1
fi
```

**Comment by BJReplay** [comment]
OK, @billraff, if you're up to it, I have a solution that it would be worthwhile to check - this installs the latest v2 Docker compose (2.26 - so roughly 17 versions newer than you're running).

If you have a look at https://piotrkrzyzek.com/install-docker-compose-plugin-v2-on-synology-dsm-7/ this sets out the instructions - but at the time this was written, the version for the instructions was the latest available: 2.17.2

The updated instructions - based on the latest release - are below:

For just your user:

``` bash
DOCKER_CONFIG=${DOCKER_CONFIG:-$HOME/.docker}
mkdir -p $DOCKER_CONFIG/cli-plugins
curl -SL https://github.com/docker/compose/releases/download/v2.26.0/docker-compose-linux-x86_64 -o $DOCKER_CONFIG/cli-plugins/docker-compose
chmod +x $DOCKER_CONFIG/cli-plugins/docker-compose
```
For all users:

``` bash
DOCKER_CONFIG=${DOCKER_CONFIG:-/usr/local/lib/docker}
mkdir -p $DOCKER_CONFIG/cli-plugins
curl -SL https://github.com/docker/compose/releases/download/v2.26.0/docker-compose-linux-x86_64 -o $DOCKER_CONFIG/cli-plugins/docker-compose
sudo chmod +x /usr/local/lib/docker/cli-plugins/docker-compose

```

To test:

`docker compose version`

Or just try again with `./upgrade.sh`

If this works, we'll work out a way to get this into the ./upgrade instructions for Synology users.

**Comment by BJReplay** [comment]
> @BJReplay I wonder if this would work:

I'm not a bash expert ;) but it looks like it would.

This is what I see on a system with docker compose V2 only - I guess you tested that on your V1 test system:

``` bash
ubuntu@mastodon:~/mastodon$ COMPOSE_VERSION=`docker-compose version --short`
Command 'docker-compose' not found, but can be installed with:
sudo snap install docker          # version 24.0.5, or
sudo apt  install docker-compose  # version 1.29.2-1
See 'snap info docker' for additional versions.
ubuntu@mastodon:~/mastodon$ COMPOSE_VERSION=`docker compose version --short`
ubuntu@mastodon:~/mastodon$ echo $COMPOSE_VERSION
2.25.0
ubuntu@mastodon:~/mastodon$
```

Interestingly, I found a WSL version that didn't have docker compose installed at all (just docker - I'd used it to build some images), so I was able to test it...

If I install a V1, COMPOSE_VERSION even with --short gives me 1.29.2

So the test might need to be > 2?

**Reply by jasonacox** [reply]
Can you create a test.sh file with my code above and run it on that V1 system and see what it says?

The bash  `[[ "${COMPOSE_VERSION}" == "2"* ]];` should only be true if the "string" starts with "2" and your example starts with "1".

**Reply by jasonacox** [reply]
Actually, since your system probably has `docker compose` (no dash) it will hit the first true and stop.  Can you check:

```bash
    if docker-compose version > /dev/null 2>&1; then
        # Check for version information to see if it is >= 2.0.0
        COMPOSE_VERSION=`docker-compose version --short`
        if [[ "${COMPOSE_VERSION}" == "2"* ]]; then
            # Build Docker (v2)
            echo "INFO: Docker Compose V2 Found: OK to Continue"
        else
            # Build Docker (v1)
            echo "ERROR: Docker Compose V1 Found: Upgrade Required"
            echo "See Migration Instructions at https://docs.docker.com/compose/migrate/"
            exit 1
        fi
```

**Comment by BJReplay** [comment]
> Actually, since your system probably has `docker compose` (no dash) it will hit the first true and stop. Can you check:

Yep, that worked on V1

**Reply by jasonacox** [reply]
@billraff I have updated the script to check for the version of docker-compose (see https://github.com/jasonacox/Powerwall-Dashboard/commit/7b9acbd43536aaa9188cafda1dcaa4b5bb68cd31). I think it should now work for you. 

Try `upgrade.sh` again and see if that works.

**Reply by billraff** [reply]
Thanks for all the work. Running the upgrade.sh now passes the docker check but the script doesn't continue:

<img width="520" alt="Screenshot 2024-03-27 at 8 05 07 AM" src="https://github.com/jasonacox/Powerwall-Dashboard/assets/29167787/5b77cf79-d497-43fc-a763-19540c3866a7">
 
An earlier question was to check the output from -

docker compose version

the result is 'compose' is not a docker command.
 I also run Powerwall-Dashboard on an old Intel Mac mini and the upgrade ran successfully there.

**Reply by jasonacox** [reply, marked as answer]
Oops! My bad. 🤦 I didn't catch the "exit" command that still existed after the docker-compose check. I just pushed a fix (see https://github.com/jasonacox/Powerwall-Dashboard/commit/cabc5849de091378b5fc5e112957df474ddfecee). You should be able too run `upgrade.sh` now and see it work. 

The good news is that it passed the check. Now let's see if the rest of it runs... Give it a try.

**Reply by billraff** [reply]
Upgrade script ran  successfully! Thanks!

**Reply by jasonacox** [reply]
Fantastic!!! Thanks @billraff - I'm sure this fix will help many others too.

And, thanks for your help @BJReplay !
